containerd-shim/kata/create.go

```diff
@@ -0,0 +1,257 @@
+	"strings"
+
+	"context"
+	vc "github.com/kata-containers/runtime/virtcontainers"
+	vf "github.com/kata-containers/runtime/virtcontainers/factory"
+	"github.com/kata-containers/runtime/virtcontainers/pkg/oci"
+	"github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/sirupsen/logrus"
+)
+
+func create(ctx context.Context, s *service, containerID, bundlePath, netns string, detach bool,
+	runtimeConfig *oci.RuntimeConfig) (vc.ContainerType, error) {
+	var err error
+
+	// Checks the MUST and MUST NOT from OCI runtime specification
+	if bundlePath, err = validCreateParams(containerID, bundlePath); err != nil {
+		return "", err
+	}
+
+	ociSpec, err := oci.ParseConfigJSON(bundlePath)
+	if err != nil {
+		return "", err
+	}
+
+	containerType, err := ociSpec.ContainerType()
+	if err != nil {
+		return "", err
+	}
+
+	//In the sandbox, the containers will only
+	//use the mnt space to separate the rootfs,
+	//and to share the other namesapces with host
+	//in the sandbox, thus remove those namespaces
+	//from ocispec except networkNamespace, since
+	//it has been ignored by kata-agent in sandbox.
+
+	for _, ns := range []specs.LinuxNamespaceType{
+		specs.UserNamespace,
+		specs.UTSNamespace,
+		specs.IPCNamespace,
+		specs.PIDNamespace,
+		specs.CgroupNamespace,
+	} {
+		removeNameSpace(&ociSpec, ns)
+	}
+
+	//set the network namespace path
+	//this set will be applied to sandbox's
+	//network config and hasn't nothing to
+	//do with containers in the sandbox since
+	//networkNamesapce hasn't been ignored by
+	//kata-agent in sandbox.
+	for _, n := range ociSpec.Linux.Namespaces {
+		if n.Type != specs.NetworkNamespace {
+			continue
+		}
+
+		if n.Path == "" {
+			n.Path = netns
+		}
+	}
+
+	setFactory(ctx, runtimeConfig)
+
+	disableOutput := noNeedForOutput(detach, ociSpec.Process.Terminal)
+
+	var c vc.VCContainer
+	switch containerType {
+	case vc.PodSandbox:
+		if s.sandbox != nil {
+			return "", fmt.Errorf("cannot create another sandbox in sandbox: %s", s.sandbox.ID())
+		}
+
+		c, err = createSandbox(ctx, ociSpec, *runtimeConfig, containerID, bundlePath, disableOutput)
+		if err != nil {
+			return "", err
+		}
+		s.sandbox = c.Sandbox()
+
+	case vc.PodContainer:
+		if s.sandbox == nil {
+			return "", fmt.Errorf("BUG: Cannot start the container, since the sandbox hasn't been created")
+		}
+
+		err = createContainer(s.sandbox, ociSpec, containerID, bundlePath, disableOutput)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	return containerType, nil
+}
+
+func setFactory(ctx context.Context, runtimeConfig *oci.RuntimeConfig) {
+	if runtimeConfig.FactoryConfig.Template {
+		factoryConfig := vf.Config{
+			Template: true,
+			VMConfig: vc.VMConfig{
+				HypervisorType:   runtimeConfig.HypervisorType,
+				HypervisorConfig: runtimeConfig.HypervisorConfig,
+				AgentType:        runtimeConfig.AgentType,
+				AgentConfig:      runtimeConfig.AgentConfig,
+			},
+		}
+		logrus.WithField("factory", factoryConfig).Info("load vm factory")
```

Comment:
This seems to be using a default logger meaning all log output will go to stderr. I don't think we want to do that? Can you  create a `logger()` function that returns a `*logrus.Entry` and leverage our logging routines to get nanosecond timestamps, journald logging, etc?

Reply:
Hi @jodh-intel 
No, the logger has been setup and redirected to a file in containerd-shim.